### PR TITLE
Implementing BiPo event data processing in REST

### PR DESCRIPTION
**Pull request by jgalan**
jgalan Large: 1105 [](https://gitlab.cern.ch/rest-for-physics/rawlib/-/commits/jgalan_bipo) [](https://github.com/rest-for-physics/rawlib/commits/jgalan_bipo)

Class documentation preview has been [rendered here](https://sultan.unizar.es/test/classTRestRawBiPoToSignalProcess.html).

In order to launch the event data processing the file needs to be uncompressed. Inside the `pipeline/external/BiPo/` there is an example of launching a basic processing chain.

We still need to uncompress the original file before start the data processing, so we need to do something like this:

```
gzip -d -c /path/to/binary/data/BiPo3Mod2_run_2600.data.gz > /tmp/BiPo3Mod2_run_2600.data
restManager --c /path/to/rml/BiPoToRawSignal.rml --f /tmp/BiPo3Mod2_run_2600.data
rm /tmp/BiPo3Mod2_run_2600.data
```

Then, we can access the information of the new generated ROOT file and quickly inspect the file using:

```
restRoot RawData_BiPo3Mod2_02600.root
```

Inside the ROOT command shell, we can access:

- Run information

```
run0->PrintMetadata();
```

Screenshot 2023-05-17 at 13 08 44

- Matacq and BiPo configuration settings:

```
md0_BiPo->PrintMetadata()
```

Screenshot 2023-05-17 at 13 10 12

- Finding some event entries that satisfy certain conditions

```
root [1] run0->GetEventEntriesWithConditions("rawAna_MaxPeakAmplitude>50 && rawAna_MaxPeakAmplitude<1000")
(std::vector<int>) { 0, 1, 2, 4, 9, 14, 17, 20, 22, 25, 26 }
```

- Updating the entry and printing analysis tree observables

```
root [3] run0->GetEntry(17)
root [4] run0->GetAnalysisTree()->PrintObservables()
```

Screenshot 2023-05-17 at 13 16 55

- Drawing the active event and saving to png

```
TCanvas c;
ev0->DrawEvent();
c.Print("Entry17.png");
```

BiPo_Run2600_Ev1714

An example was added to include the analysis (using TRestRawSignalAnalysisProcess) of two different range of the waveforms, the example at `examples/01.BiPo/` contains a `plots.rml` definition that will generate few prebuilt plots.

```
restManager --c plots.rml --f RawData_BiPo3Mod2_02600.root
```

R02503_BiPo3Mod2_Raw

### CODE REVIEW

File: inc/TRestRawBiPoToSignalProcess.h [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+ * For the list of contributors see $REST_PATH/CREDITS.                  *
+ *************************************************************************/
+
+#ifndef RestCore_TRestBiPoToSignalProcess
+#define RestCore_TRestBiPoToSignalProcess
+
+#include "TRestRawSignalEvent.h"
+#include "TRestRawToSignalProcess.h"
+
+static const size_t CTAG_SZ = 4;
```
Comment (juanangp) [inline inc/TRestRawBiPoToSignalProcess.h:29]:
```suggestion
static constexpr size_t CTAG_SZ = 4;
```

File: inc/TRestRawBiPoToSignalProcess.h [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+static const int MATACQ_UNDERFLOW = 0x0000;
+static const int MATACQ_ZERO = 0x8000;
+static const int MATACQ_OVERFLOW = 0xFFFF;
+
+const std::string TAG_RUN_START = "STA";
+const std::string TAG_RUN_BIPO = "ST2";
+const std::string TAG_RUN_STOP = "STO";
+const std::string TAG_ACQ = "ACQ";
+const std::string TAG_ACQ_2 = "AC2";
```
Comment (juanangp) [inline inc/TRestRawBiPoToSignalProcess.h:43]:
```suggestion
constexpr int MATACQ_N_CH = 4;               // Number of channels
constexpr int MATACQ_MAX_DATA_SAMP = 10240;  // Max number of samples for
constexpr int MATACQ_BIPO_TIMEOUT = 0;       // BiPo mode timeout
static constexpr int MATACQ_UNDERFLOW = 0x0000;
static constexpr int MATACQ_ZERO = 0x8000;
static constexpr int MATACQ_OVERFLOW = 0xFFFF;

constexpr std::string TAG_RUN_START = "STA";
constexpr std::string TAG_RUN_BIPO = "ST2";
constexpr std::string TAG_RUN_STOP = "STO";
constexpr std::string TAG_ACQ = "ACQ";
constexpr std::string TAG_ACQ_2 = "AC2";
```

File: inc/TRestRawBiPoToSignalProcess.h [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+    int32_t Nb_Acq;
+
+    int32_t Posttrig;
+
+    int32_t Time_Tag_On;
+
+    int32_t Sampling_GHz;
+
+    int32_t ch_shifts[MATACQ_N_CH];
```
Comment (juanangp) [inline inc/TRestRawBiPoToSignalProcess.h:65]:
```suggestion
    std::array <int32_t, MATACQ_N_CH>ch_shifts;
```

File: inc/TRestRawBiPoToSignalProcess.h [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+/// A structure to store the BiPo settings
+struct BiPoSettings {
+    int32_t trigger_address;
+
+    int32_t Win1_Posttrig;
+    int32_t Timeout_200KHz;
+
+    int32_t Trig_Chan[MATACQ_N_CH];
+    int32_t Level1_mV[MATACQ_N_CH];
+    int32_t Level2_mV[MATACQ_N_CH];
```
Comment (juanangp) [inline inc/TRestRawBiPoToSignalProcess.h:79]:
```suggestion
    std::array <int32_t, MATACQ_N_CH> Trig_Chan;
    std::array <int32_t, MATACQ_N_CH> Level1_mV;
    std::array <int32_t, MATACQ_N_CH> Level2_mV;
```

File: src/TRestRawBiPoToSignalProcess.cxx [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+/// \brief This method reads the event data corresponding to one event.
+/// The sampled channel data that will be made accessible at the `mdata`
+/// pointer provided as argument.
+///
+/// The event timestamp and the triggered board values will be also read
+/// here. The event timestamp will be assigned to the fSignalEvent, while
+/// the triggered board address will be returned and it will be used
+/// later on to generate a signal id.
+///
+Int_t TRestRawBiPoToSignalProcess::ReadBiPoEventData(uint16_t* mdata) {
```
Comment (juanangp) [inline src/TRestRawBiPoToSignalProcess.cxx:458]: See above
```suggestion
Int_t TRestRawBiPoToSignalProcess::ReadBiPoEventData(std::vector <uint16_t>& mdata) {
```

File: src/TRestRawBiPoToSignalProcess.cxx [hunk 6]
```diff
@@ -0,0 +1,573 @@
...
+    RESTDebug << "Data size --> " << tmp << RESTendl;
+
+    if (fread(&tmp, sizeof(int32_t), 1, fInputBinFile) != 1) {
+        printf("Error: could not read BiPo trigger address.\n");
+        exit(1);
+    }
+    totalBytesReaded += sizeof(int32_t);
+    RESTDebug << " T1-T2 distance --> " << tmp << RESTendl;
+
+    if (fread(mdata, sizeof(uint16_t), data_size, fInputBinFile) != (size_t)data_size) {
```
Comment (juanangp) [inline src/TRestRawBiPoToSignalProcess.cxx:506]:
```suggestion
    mdata.resize(data_size);
    if (fread(&mdata[0], sizeof(uint16_t), data_size, fInputBinFile) != (size_t)data_size) {
```

**Comment by jgalan** [comment]
@juanangp I get the following compilation error.

```
error: constexpr variable cannot have non-literal type 'const std::string' (aka 'const basic_string<char>')
constexpr std::string TAG_ACQ_2 = "AC2";
```

**Comment by jgalan** [comment]
Then I get plenty of errors, because `fread` requires a `void *` and changing it to `std::array` the compilation complains.

**Comment by juanangp** [comment]
> @juanangp I get the following compilation error.
> 
> ```
> error: constexpr variable cannot have non-literal type 'const std::string' (aka 'const basic_string<char>')
> constexpr std::string TAG_ACQ_2 = "AC2";
> ```
> 
> what is the problem using just `const`?

Indeed, `constexpr` cannot be used with `std::string`, but should be possible to do `constexpr std::string_view TAG_ACQ_2 = "AC2";` or `constexpr char[] TAG_ACQ_2 = "AC2";`

Is better to use `constexpr` in this case because it would be already available at compile time, which means that will be faster https://stackoverflow.com/questions/41125651/constexpr-vs-static-const-which-one-to-prefer. Anyhow, it is not a big deal but I think is a good practise to use  `constexpr`.

**Comment by juanangp** [comment]
> Then I get plenty of errors, because `fread` requires a `void *` and changing it to `std::array` the compilation complains. So I just pushed force the original version.
> 
> Perhaps this is a change to be made once the validation pipeline is running for this example.

I guess you have to do something like `&stdArray[0]`, I can have a look later on.

**Review by Anakintana — APPROVED** [review]

**Review by juanangp — APPROVED** [review]

**Pull request merged**